### ISSUE TITLE
[spike] Investigate our Theming Implementation

### ISSUE THREAD

**Issue by JakeLaCombe** [cerner/terra-core#723]
# Issue Description
We currently have two approaches that we are looking into with doing our theming approach. There exist a [theme provider](https://github.com/cerner/terra-core/blob/theme-adapter/packages/terra-theme-provider/src/CSSVariableProvider.jsx#L29) branch that takes in a json file of attributes, and uses JavaScript to manually set the styles inside the current CSS. Another approach being investigated is using classes to set the variable names and swapping classes on the theming component to update the values. This approach can be viewed [here](https://www.google.com/search?q=theming+with+css+variables&rlz=1C5CHFA_enUS719US722&oq=theming+with+css+variables&aqs=chrome..69i57j69i60.4583j0j7&sourceid=chrome&ie=UTF-8)

The goal of this spike is to determine what we would like to do in order to implement theming. While we have talked about the two approaches logged up above, our team is open to other approaches that meets our needs as well.

We also should document what browser support we will have to dynamic theming as well. Older browsers don't support css variables, which can make some of our suggested implementations invalid.

**Comment by mhemesath** [comment]
I see a few ways to achieve this.

1. Each component contains all the 1st class themes.. This makes the theme provider do nothing more than flip a class name to activate a theme. To keep theme classes consistent.. we could use SASS mixins to ensure each component scopes the theme consistenly. e.g. The SASS mixin 

```css
// theme-mixins.sass
$default-theme: false !default
$bundled-themes: "consumer" "ocs"; !default

@mixin consumer-theme {
  @if index('consumer, $bundled-themes) {
    @if $default-theme == "consumer"
      .terra-consumer-theme {
        @content 
      }
    } @else {
      :root {
        @content 
      }
    }
  }  
}

@mixin ocs-theme {
  @if index('ocs, $bundled-themes) {
    @if $default-theme == "ocs"
      .terra-ocs-theme {
        @content 
      }
    } @else {
      :root {
        @content 
      }
    }
  } 
}


//=======================
// badge.sass

@include consumer-theme {
  --color: purple;
}

@include ocs-theme {
  --color: blue;
}

.badge {
  color: var(--color, black);
}
```
We can then set a sass variable in webpack to set default and bundled themes:

```js
{
    loader: "sass-loader",
    options: {
        data: "$default-theme: consumer; $bundled-themes: consumer"
    }
}
```

**Comment by mhemesath** [comment]
Its also possilbe we could do something similar above with postcss. It could be done via a few strategies:

1.  Creating a custom theming selector .e.g `:theme(consumer) { }`
2. Enabling postcss mixins/vars plugins to achieve similar strategy above.

The downside with postCSS is it adds complexity to that plugin setup. One solution to that would be create  terra-postCSS wrapper that would bundle all required config to make that easier for webpack integration.

**Comment by zhongyn** [comment]
With the SASS solution, it will require a new mixin in theme-mixins.scss whenever a new theme is added. Can we use a higher level mixin? so that we can add a theme name in one place in theme-mixins.scss like:
```
theme-map: {
   consumer: terra-consumer-theme;
   ocs: terra-ocs-theme;
}
```
And the higher level mixin will generate all required submixins?

**Comment by mhemesath** [comment]
>  it will require a new mixin in theme-mixins.scss whenever a new theme is added

We shouldn't really have that many new themes. Terra-core really should only be expected to support first-class themes IMO

**Comment by zhongyn** [comment]
how do we define first-class themes? How does the above solution work with the scenario that a badge has multiple consumer theme colors? (like button with different concept colors)

**Comment by zhongyn** [comment]
a bit confused, in one of the comments here https://github.com/cerner/terra-core/issues/729, it is mentioned that not to scope sub concept themes in consumer theme

**Comment by mhemesath** [comment]
I know... I changed my mind based on how we're probably going to do theme packaging

**Comment by bjankord** [comment]
+1 on theme mixins approach. I like how this makes it easy for us to control first class themes, and user can opt into the themes they need. Some of the logic in the mixin needed flipped, I coded up a proof on concept of this and sent out a PR here: https://github.com/cerner/terra-core/pull/732

This PR is just for review and to further discuss this approach.

In the PR, I've set `consumer` as the default theme. All browsers (IE10/IE11 included) will display badges with the `consumer` theme by default. Users in modern browsers get a theme switcher at the top of the site, and can switch to a mock `OCS` theme.
